src/codegen/cpu.rs

```diff
@@ -129,6 +87,101 @@ impl<'a> X86CodeGenerator<'a> {
     }
 
+    pub fn emit_main<W: Write>(
+        &mut self,
+        top_arg_tensors: &'a [Rc<Tensor<Tgt>>],
+        out: &mut W,
+    ) -> fmt::Result {
+        let mut main_body_str = String::new();
+        let mut depth = 0_usize;
+
+        writeln!(main_body_str, "int main() {{")?;
+        depth += 1;
+        for (i, tensor) in top_arg_tensors.iter().enumerate() {
+            let c_buffer = self.name_env.get(tensor).unwrap();
+            writeln!(
+                main_body_str,
+                "{}{} buf{}[{}] __attribute__((aligned (128)));",
+                indent(depth),
+                c_type(c_buffer.dtype()),
+                i, // we do not use c_buffer.name().unwrap()
+                c_buffer.size().unwrap(),
+            )?;
+        }
+        writeln!(main_body_str)?;
+
+        // Initialize the input tensors
+        writeln!(
+            main_body_str,
+            "{}// Initialize the input tensors.",
+            indent(depth)
+        )?;
+        for (i, tensor) in top_arg_tensors
+            .iter()
+            .enumerate()
+            .take(top_arg_tensors.len() - 1)
+        {
+            let c_buffer = self.name_env.get(tensor).unwrap();
+            writeln!(
+                main_body_str,
+                "{}for (int i = 0; i < {}; i++) {{",
+                indent(depth),
+                c_buffer.size().unwrap(),
+            )?;
+            depth += 1;
+            writeln!(
+                main_body_str,
+                "{}buf{}[i] = i;",
+                indent(depth),
+                i, // we do not use c_buffer.name().unwrap()
+            )?;
+            depth -= 1;
+            writeln!(main_body_str, "{}}}", indent(depth))?;
+        }
+        writeln!(main_body_str)?;
+
+        // Emit the kernel call
+        write!(main_body_str, "{}kernel(", indent(depth))?;
+        for i in 0..top_arg_tensors.len() {
+            write!(
+                main_body_str,
+                "&buf{}[0]{}",
+                i, // we do not use c_buffer.name().unwrap()
+                if i < top_arg_tensors.len() - 1 {
+                    ", "
+                } else {
+                    ");\n"
+                },
+            )?;
+        }
+        writeln!(main_body_str)?;
+
+        // Print the output tensor
+        writeln!(main_body_str, "{}// Print the output.", indent(depth))?;
+        let tensor = &top_arg_tensors[top_arg_tensors.len() - 1];
+        let c_buffer = self.name_env.get(tensor).unwrap();
+        writeln!(
+            main_body_str,
+            "{}for (int i = 0; i < {}; i++) {{",
+            indent(depth),
+            c_buffer.size().unwrap(),
+        )?;
+        depth += 1;
+        writeln!(
+            main_body_str,
+            "{}printf(\"%d \", buf{}[i]);",
+            indent(depth),
+            top_arg_tensors.len() - 1, // we do not use c_buffer.name().unwrap()
+        )?;
+        depth -= 1;
+        writeln!(main_body_str, "{}}}", indent(depth))?;
```

Comment:
It looks like this differs from the Python implementation in that it prints all the values in memory order. The Python implementation adds newlines to display shape and prints values in logical order (independent of memory layout). If this is right, let's add a TODO noting this. (We'll need to fix it soon.)

Reply:
Sorry, could you elaborate on what you meant here? I followed this C code generated by the Python implementation.

```c
int main() {
  uint32_t buf0[32] __attribute__((aligned (128)));
  uint32_t buf1[32] __attribute__((aligned (128)));
  uint32_t buf2[64] __attribute__((aligned (128)));

  // Initialize the input data.
    for (int i = 0; i < 32; ++i) {
        buf0[i] = i;
        buf1[i] = i;
    }

  kernel(&buf0[0], &buf1[0], &buf2[0]);

  // Print the output.
    for (int i = 0; i < 64; ++i) {
        printf("%d ", buf2[i]);
    }

  return 0;
}
```

Reply:
Here's the output of a size-2 Matmul from the Python implementation:

```c
#include <inttypes.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <immintrin.h>
#define is_aligned(POINTER, BYTE_COUNT) \
  (((uintptr_t)(const void *)(POINTER)) % (BYTE_COUNT) == 0)
typedef uint32_t vui8 __attribute__ ((vector_size (32)));
typedef uint32_t vui4 __attribute__ ((vector_size (16)));
typedef uint8_t vub32 __attribute__ ((vector_size (32)));
typedef uint8_t vub16 __attribute__ ((vector_size (16)));
__attribute__((noinline))
void kernel(
  uint32_t *restrict buf0, 
  uint32_t *restrict buf1, 
  uint32_t *restrict buf2)
{
  uint32_t buf3[4] __attribute__((aligned (128)));        // Matmul((2×2, u32, L1), (2×2, u32), (2×2, u32), serial)
  for (int t0 = 0; t0 < 2; t0++) {        // Load((2×2, u32, L1), (2×2, u32, RF), serial)
  for (int t1 = 0; t1 < 2; t1++) {
    buf3[(t1 + (2 * t0))] = buf0[(t1 + (2 * t0))];  // ValueAssign        // Load((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial)
  }
  }
  uint32_t buf4[4] __attribute__((aligned (128)));        // Matmul((2×2, u32, RF), (2×2, u32, L1), (2×2, u32), serial)
  for (int t2 = 0; t2 < 2; t2++) {        // Load((2×2, u32, L1), (2×2, u32, RF), serial)
  for (int t3 = 0; t3 < 2; t3++) {
    buf4[(t3 + (2 * t2))] = buf1[(t3 + (2 * t2))];  // ValueAssign        // Load((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial)
  }
  }
  uint32_t buf5[4] __attribute__((aligned (128)));        // Matmul((2×2, u32, RF), (2×2, u32, RF), (2×2, u32, L1), serial)
  memset((void *)(&buf5[0]), 0, 16);  /* MemsetZero */        // Zero((2×2, u32, RF), serial)
  for (int t4 = 0; t4 < 2; t4++) {        // MatmulAccum((2×2, u32, RF), (2×2, u32, RF), (2×2, u32, RF), serial)
  for (int t5 = 0; t5 < 2; t5++) {
    for (int t6 = 0; t6 < 2; t6++) {        // MatmulAccum((1×2, u32, RF, ua), (2×1, u32, RF, c1, ua), (1×1, u32, RF, ua), serial)
      buf5[(t5 + (2 * t4))] += buf3[(t6 + (2 * t4))] * buf4[(t5 + (2 * t6))];  /* Mult */        // MatmulAccum((1×1, u32, RF, ua), (1×1, u32, RF, ua), (1×1, u32, RF, ua), serial)
    }
  }
  }
  for (int t7 = 0; t7 < 2; t7++) {        // Store((2×2, u32, L1), (2×2, u32, RF), serial)
  for (int t8 = 0; t8 < 2; t8++) {
    buf2[(t8 + (2 * t7))] = buf5[(t8 + (2 * t7))];  // ValueAssign (store)        // Store((1×1, u32, L1, ua), (1×1, u32, RF, ua), serial)
  }
  }
}

int main() {
  uint32_t buf0[4] __attribute__((aligned (128)));
  uint32_t buf1[4] __attribute__((aligned (128)));
  uint32_t buf2[4] __attribute__((aligned (128)));
  kernel(&buf0[0], &buf1[0], &buf2[0]);
  // Print buf2
  {
    printf("2x2\n");
    for (size_t a = 0; a < 2; a++) {
    for (size_t b = 0; b < 2; b++) {
      printf("%" PRIu32 " ", buf2[(b + (2 * a))]);
    }
    printf("\n");
    }
  }
  return 0;
}
```

Reply:
What TODO comment would you recommend adding here?

Reply:
To put a TODO comment here,  what comment would you recommend adding here?

Reply:
Instead, I implemented this in [46d2e89](https://github.com/samkaufman/morello/pull/10/commits/46d2e89c1187f406a7647ed9206bbab1d96e96ff).